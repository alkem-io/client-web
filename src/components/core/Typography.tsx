import clsx from 'clsx';
import React, { FC, useLayoutEffect, useMemo, useRef } from 'react';
import { Palette, Typography as TypographyContract } from '../../themes';
import { createStyles } from '../../hooks/useTheme';
import _clamp from 'clamp-js';
import { replaceAll } from '../../utils/replaceAll';

const useTypographyStyles = createStyles(theme => ({
  h1: {
    fontFamily: theme.typography.h1.font,
    fontSize: theme.typography.h1.size,
  },
  h2: {
    fontFamily: theme.typography.h2.font,
    fontSize: theme.typography.h2.size,
    textTransform: 'uppercase',
    marginBottom: theme.shape.spacing(1),
  },
  h3: {
    fontFamily: theme.typography.h3.font,
    fontSize: theme.typography.h3.size,
    marginBottom: theme.shape.spacing(1),
  },
  h4: {
    fontFamily: theme.typography.h4.font,
    fontSize: theme.typography.h4.size,
    marginBottom: theme.shape.spacing(1),
  },
  h5: {
    fontFamily: theme.typography.h5.font,
    fontSize: theme.typography.h5.size,
  },
  caption: {
    fontFamily: theme.typography.caption.font,
    fontSize: theme.typography.caption.size,
    textTransform: 'uppercase',
  },
  body: {
    fontFamily: theme.typography.body.font,
    fontSize: theme.typography.body.size,
  },
  button: {
    fontFamily: theme.typography.button.font,
    fontSize: theme.typography.button.size,
    textTransform: 'uppercase',
  },
  primary: {
    color: theme.palette.primary,
  },
  positive: {
    color: theme.palette.positive,
  },
  neutral: {
    color: theme.palette.neutral,
  },
  neutralMedium: {
    color: theme.palette.neutralMedium,
  },
  neutralLight: {
    color: theme.palette.neutralLight,
  },
  negative: {
    color: theme.palette.negative,
  },
  divider: {
    color: theme.palette.divider,
  },
  background: {
    color: theme.palette.background,
  },
  inherit: {
    color: 'inherit',
  },
}));

interface FontWeight {
  regular: number;
  medium: number;
  boldLight: number;
  bold: number;
}

const fontWeight: FontWeight = {
  regular: 400,
  medium: 500,
  boldLight: 700,
  bold: 900,
};

interface TypographyProps extends Record<string, unknown> {
  variant?: keyof TypographyContract;
  className?: string;
  color?: keyof Palette | 'inherit';
  weight?: keyof FontWeight;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  as?: React.ComponentType<any> | string;
  classes?: unknown;
  clamp?: string | number;
}

const Typography: FC<TypographyProps> = ({
  variant = 'body',
  color = 'neutral',
  weight = 'medium',
  as: Component = 'div',
  className,
  classes,
  children,
  clamp = undefined,
  ...rest
}) => {
  const styles = useTypographyStyles(classes);
  const ref = useRef();

  useLayoutEffect(() => {
    if (clamp && ref?.current) {
      _clamp(ref.current, { clamp });
    }
  });

  const removeSlashes = (string: string) => replaceAll('\\"', '"', replaceAll('\\n', '<br/>', string));

  const plainText = children as string;

  const text = useMemo(() => {
    if (children instanceof Array) {
      return children.map(c => {
        if (typeof c === 'string') return <span key={c} dangerouslySetInnerHTML={{ __html: removeSlashes(c) }} />;
        return c;
      });
    }
    if (typeof children === 'string') {
      return <span dangerouslySetInnerHTML={{ __html: removeSlashes(plainText) }} />;
    }
    return children;
  }, [children]);

  return (
    <Component
      {...rest}
      className={clsx(styles[variant], styles[color], className)}
      style={{ fontWeight: fontWeight[weight] }}
      ref={ref}
    >
      {text}
    </Component>
  );
};

export default Typography;
