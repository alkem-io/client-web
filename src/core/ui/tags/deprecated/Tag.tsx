import { Typography } from '@mui/material';
import clsx from 'clsx';
import React from 'react';
import { makeStyles } from '@mui/styles';

const useTagStyles = makeStyles(theme => ({
  tag: {
    padding: `${theme.spacing(0.7)} ${theme.spacing(1.4)}`,
    display: 'inline-flex',
    maxWidth: '100%',
  },
  primary: {
    background: theme.palette.primary.main,
    color: theme.palette.neutralLight.main,
  },
  positive: {
    background: theme.palette.positive.main,
    color: theme.palette.neutralLight.main,
  },
  neutralMedium: {
    background: theme.palette.neutralMedium.main,
    color: theme.palette.background.paper,
  },
  neutral: {
    background: theme.palette.neutral.main,
    color: theme.palette.background.paper,
  },
  negative: {
    background: theme.palette.negative.main,
    color: theme.palette.background.paper,
  },
  background: {
    background: theme.palette.background.paper,
    color: theme.palette.primary.main,
    border: `1px solid ${theme.palette.primary.main}`,
  },
}));

export interface TagProps extends React.SVGProps<SVGSVGElement> {
  color?: 'positive' | 'neutralMedium' | 'primary' | 'neutral' | 'negative' | 'background'; //keyof Palette
  text: string | React.ReactNode;
  className?: string;
  actions?: number | undefined;
}

/**
 * @deprecated
 */
const Tag: React.FC<TagProps> = ({ text, className, color = 'positive' }): JSX.Element | null => {
  const styles = useTagStyles();

  return (
    <div className={clsx(styles.tag, styles[color], className)}>
      <Typography display={'block'} variant={'caption'} noWrap>
        {text}
      </Typography>
    </div>
  );
};

export default Tag;
