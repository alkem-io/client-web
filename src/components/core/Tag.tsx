import clsx from 'clsx';
import React from 'react';
import { createStyles } from '../../hooks/useTheme';
import Typography from './Typography';

const useIconStyles = createStyles(theme => ({
  tag: {
    padding: `${theme.shape.spacing(0.7)}px ${theme.shape.spacing(1.4)}px`,
    display: 'inline-flex',
  },
  primary: {
    background: theme.palette.primary,
    color: theme.palette.neutralLight,
  },
  positive: {
    background: theme.palette.positive,
    color: theme.palette.neutralLight,
  },
  neutralMedium: {
    background: theme.palette.neutralMedium,
    color: theme.palette.background,
  },
  neutral: {
    background: theme.palette.neutral,
    color: theme.palette.background,
  },
  negative: {
    background: theme.palette.negative,
    color: theme.palette.background,
  },
  background: {
    background: theme.palette.background,
    color: theme.palette.primary,
    border: `1px solid ${theme.palette.primary}`,
  },
}));

export interface TagProps extends React.SVGProps<SVGSVGElement> {
  color?: 'positive' | 'neutralMedium' | 'primary' | 'neutral' | 'negative' | 'background'; //keyof Palette
  text: string;
  className?: string;
}

const Tag: React.FC<TagProps> = ({ text, className, color = 'positive' }): JSX.Element | null => {
  const styles = useIconStyles();

  return (
    <span className={clsx(styles.tag, styles[color], className)}>
      <Typography variant="caption" color="inherit" clamp={1}>
        {text}
      </Typography>
    </span>
  );
};

export default Tag;
