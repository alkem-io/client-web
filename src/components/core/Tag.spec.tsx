import renderer from 'react-test-renderer';
import React from 'react';
import Tag from './Tag';
import Typography from './Typography';
import { render, screen } from '@testing-library/react';

describe('Tag component', () => {
  test('render correctly Typography component', () => {
    // arrange
    const { asFragment } = render(<Typography variant="caption" color="inherit" />);

    // act
    const html = asFragment();

    // assert
    expect(html).toMatchSnapshot();
  });

  test('render correctly Tag component', () => {
    // arrange
    const text = 'Please wait';

    // act
    const TextComponent = renderer.create(<Tag text={text} />).toJSON();

    // assert
    expect(TextComponent).toMatchSnapshot();
  });

  test('check Tag with message', async () => {
    // arrange
    const message = 'Tag message';

    // act
    render(<Tag text={message} />);

    // assert
    expect(screen.getByText(message)).toBeInTheDocument();
  });
});
