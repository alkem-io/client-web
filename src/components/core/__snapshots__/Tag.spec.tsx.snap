// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly Tag component 1`] = `
<span
  className="tag-0-2-18 positive-0-2-20"
>
  <div
    className="caption-0-2-30 inherit-0-2-41"
    style={
      Object {
        "fontWeight": 500,
      }
    }
  >
    <span
      dangerouslySetInnerHTML={
        Object {
          "__html": "Please wait",
        }
      }
    />
  </div>
</span>
`;

exports[`render correctly Typography component 1`] = `ReactWrapper {}`;


